<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.134">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>ANOVA</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/cs2_practical_kruskal-wallis.html" rel="next">
<link href="../materials/cs1_practical_two-sample-paired-wilcoxon.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><script src="../site_libs/kePrint-0.0.1/kePrint.js"></script><link href="../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo"></a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../setup.html">Setup</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../materials.html" aria-current="page">Materials</a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/cambiotraining/corestats/"><i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs"><i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
 </a>
  </li>  
</ul>
<div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">ANOVA</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials.html" class="sidebar-item-text sidebar-link">Materials</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Simple hypothesis testing</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_one-sample-t-test.html" class="sidebar-item-text sidebar-link">One-sample t-test</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_one-sample-wilcoxon.html" class="sidebar-item-text sidebar-link">Wilcoxon signed-rank test</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_two-sample-t-test.html" class="sidebar-item-text sidebar-link">Student’s t-test</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_two-sample-mann-whitney.html" class="sidebar-item-text sidebar-link">Mann-Whitney U test</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_two-sample-paired-t-test.html" class="sidebar-item-text sidebar-link">Paired t-test</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs1_practical_two-sample-paired-wilcoxon.html" class="sidebar-item-text sidebar-link">Wilcoxon signed rank test</a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Categorical predictors</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs2_practical_anova.html" class="sidebar-item-text sidebar-link active">ANOVA</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs2_practical_kruskal-wallis.html" class="sidebar-item-text sidebar-link">Kruskal-Wallis</a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Continuous predictors</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs3_practical_linear-regression.html" class="sidebar-item-text sidebar-link">Linear regression</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/cs3_practical_correlations.html" class="sidebar-item-text sidebar-link">Correlations</a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#purpose-and-aim" id="toc-purpose-and-aim" class="nav-link active" data-scroll-target="#purpose-and-aim">Purpose and aim</a></li>
  <li><a href="#libraries-and-functions" id="toc-libraries-and-functions" class="nav-link" data-scroll-target="#libraries-and-functions">Libraries and functions</a></li>
  <li><a href="#data-and-hypotheses" id="toc-data-and-hypotheses" class="nav-link" data-scroll-target="#data-and-hypotheses">Data and hypotheses</a></li>
  <li><a href="#summarise-and-visualise" id="toc-summarise-and-visualise" class="nav-link" data-scroll-target="#summarise-and-visualise">Summarise and visualise</a></li>
  <li>
<a href="#assumptions" id="toc-assumptions" class="nav-link" data-scroll-target="#assumptions">Assumptions</a>
  <ul class="collapse">
<li><a href="#normality" id="toc-normality" class="nav-link" data-scroll-target="#normality">Normality</a></li>
  <li><a href="#equality-of-variance" id="toc-equality-of-variance" class="nav-link" data-scroll-target="#equality-of-variance">Equality of Variance</a></li>
  <li><a href="#graphical-interpretation-and-diagnostic-plots" id="toc-graphical-interpretation-and-diagnostic-plots" class="nav-link" data-scroll-target="#graphical-interpretation-and-diagnostic-plots">Graphical interpretation and diagnostic plots</a></li>
  </ul>
</li>
  <li><a href="#implement-and-interpret-the-test" id="toc-implement-and-interpret-the-test" class="nav-link" data-scroll-target="#implement-and-interpret-the-test">Implement and interpret the test</a></li>
  <li>
<a href="#post-hoc-testing-tukeys-rank-test" id="toc-post-hoc-testing-tukeys-rank-test" class="nav-link" data-scroll-target="#post-hoc-testing-tukeys-rank-test">Post-hoc testing (Tukey’s rank test)</a>
  <ul class="collapse">
<li><a href="#read-in-data-and-plot" id="toc-read-in-data-and-plot" class="nav-link" data-scroll-target="#read-in-data-and-plot">Read in data and plot</a></li>
  <li><a href="#test-for-a-significant-difference-in-group-means" id="toc-test-for-a-significant-difference-in-group-means" class="nav-link" data-scroll-target="#test-for-a-significant-difference-in-group-means">Test for a significant difference in group means</a></li>
  <li><a href="#performing-tukeys-test" id="toc-performing-tukeys-test" class="nav-link" data-scroll-target="#performing-tukeys-test">Performing Tukey’s test</a></li>
  <li><a href="#assumptions-1" id="toc-assumptions-1" class="nav-link" data-scroll-target="#assumptions-1">Assumptions</a></li>
  </ul>
</li>
  <li><a href="#exercise-lobster-weight" id="toc-exercise-lobster-weight" class="nav-link" data-scroll-target="#exercise-lobster-weight">Exercise: Lobster weight</a></li>
  <li><a href="#key-points" id="toc-key-points" class="nav-link" data-scroll-target="#key-points">Key points</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">ANOVA</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header><div class="cell">

</div>
<div class="cell">

</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning outcomes
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Questions</strong></p>
<ul>
<li>How do I analyse multiple samples of continuous data?</li>
<li>What is an ANOVA?</li>
<li>How do I check for differences between groups?</li>
</ul>
<p><strong>Objectives</strong></p>
<ul>
<li>Be able to perform an ANOVA in R</li>
<li>Understand the ANOVA output and evaluate the assumptions</li>
<li>Understand what post-hoc testing is and how to do this in R</li>
</ul>
</div>
</div>
<section id="purpose-and-aim" class="level2"><h2 class="anchored" data-anchor-id="purpose-and-aim">Purpose and aim</h2>
<p>Analysis of variance or ANOVA is a test than can be used when we have multiple samples of continuous response data. Whilst it is possible to use ANOVA with only two samples, it is generally used when we have three or more groups. It is used to find out if the samples came from parent distributions with the same mean. It can be thought of as a generalisation of the two-sample Student’s t-test.</p>
</section><section id="libraries-and-functions" class="level2"><h2 class="anchored" data-anchor-id="libraries-and-functions">Libraries and functions</h2>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click to expand
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<section id="libraries" class="level3"><h3 class="anchored" data-anchor-id="libraries">Libraries</h3>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># A collection of R packages designed for data science</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Converts stats functions to a tidyverse-friendly format</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/rstatix/">rstatix</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Creates diagnostic plots using ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://goodekat.github.io/ggResidpanel/">ggResidpanel</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="functions" class="level3"><h3 class="anchored" data-anchor-id="functions">Functions</h3>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Computes summary statistics</span></span>
<span><span class="fu">rstatix</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/get_summary_stats.html">get_summary_stats</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Creates diagnostic plots</span></span>
<span><span class="fu">ggResidpanel</span><span class="fu">::</span><span class="fu"><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">resid_panel</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<section id="functions-1" class="level3"><h3 class="anchored" data-anchor-id="functions-1">Functions</h3>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Fits a linear model  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Carries out an ANOVA on a linear model </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plots a Q-Q plot for comparison with a normal distribution</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html">qqnorm</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adds a comparison line to the Q-Q plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html">qqline</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Performs a Shapiro-Wilk test for normality</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<table class="table">
<thead><tr class="header">
<th style="text-align: left;">Libraries</th>
<th style="text-align: left;">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/getting_started/index.html"><code>pandas</code></a></td>
<td style="text-align: left;">A Python data analysis and manipulation tool.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://pingouin-stats.org"><code>pingouin</code></a></td>
<td style="text-align: left;">A Python module developed to have simple yet exhaustive stats functions.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://plotnine.readthedocs.io/en/stable/"><code>plotnine</code></a></td>
<td style="text-align: left;">The Python equivalent of <code>ggplot2</code>.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://www.statsmodels.org/stable/index.html"><code>statsmodels</code></a></td>
<td style="text-align: left;">A Python module for statistical models, conducting tests and statistical data exploration.</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th style="text-align: left;">Functions</th>
<th style="text-align: left;">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html"><code>pandas.DataFrame.describe()</code></a></td>
<td style="text-align: left;">Gives summary statistics.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html"><code>pandas.DataFrame.groupby()</code></a></td>
<td style="text-align: left;">Group DataFrame using a mapper or by a Series of columns</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html"><code>pandas.DataFrame.head()</code></a></td>
<td style="text-align: left;">Plots the first few rows of a DataFrame.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html"><code>pandas.DataFrame.query()</code></a></td>
<td style="text-align: left;">Query the columns of a DataFrame with a boolean expression</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html"><code>pandas.read_csv</code></a></td>
<td style="text-align: left;">Reads in a <code>.csv</code> file.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://pingouin-stats.org/generated/pingouin.anova.html#pingouin.anova"><code>pingouin.anova()</code></a></td>
<td style="text-align: left;">Performs an analysis of variance.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://pingouin-stats.org/generated/pingouin.homoscedasticity.html#pingouin.homoscedasticity"><code>pingouin.homoscedasticity()</code></a></td>
<td style="text-align: left;">Tests equality of variance.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://pingouin-stats.org/generated/pingouin.normality.html"><code>pingouin.normality()</code></a></td>
<td style="text-align: left;">Performs the Shapiro-Wilk test for normality.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://www.statsmodels.org/stable/generated/statsmodels.formula.api.ols.html#statsmodels.formula.api.ols"><code>statsmodels.formula.api.ols</code></a></td>
<td style="text-align: left;">Creates a model from a formula and data frame.</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="data-and-hypotheses" class="level2"><h2 class="anchored" data-anchor-id="data-and-hypotheses">Data and hypotheses</h2>
<p>For example, suppose we measure the feeding rate of oyster catchers (shellfish per hour) at three sites characterised by their degree of shelter from the wind, imaginatively called <code>exposed</code> (E), <code>partially sheltered</code> (P) and <code>sheltered</code> (S). We want to test whether the data support the hypothesis that feeding rates don’t differ between locations. We form the following null and alternative hypotheses:</p>
<ul>
<li>
<span class="math inline">\(H_0\)</span>: The mean feeding rates at all three sites is the same <span class="math inline">\(\mu E = \mu P = \mu S\)</span>
</li>
<li>
<span class="math inline">\(H_1\)</span>: The mean feeding rates are not all equal.</li>
</ul>
<p>We will use a one-way ANOVA test to check this.</p>
<ul>
<li>We use a <strong>one-way</strong> ANOVA test because we only have one predictor variable (the categorical variable location).</li>
<li>We’re using <strong>ANOVA</strong> because we have more than two groups and we don’t know any better yet with respect to the exact assumptions.</li>
</ul>
<p>The data are stored in the file <code>data/CS2-oystercatcher.csv</code>.</p>
</section><section id="summarise-and-visualise" class="level2"><h2 class="anchored" data-anchor-id="summarise-and-visualise">Summarise and visualise</h2>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<p>First we read in the data.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load data</span></span>
<span><span class="va">oystercatcher</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/CS2-oystercatcher-feeding.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># and have a look</span></span>
<span><span class="va">oystercatcher</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 120 × 2
   site    feeding
   &lt;chr&gt;     &lt;dbl&gt;
 1 exposed    12.2
 2 exposed    13.1
 3 exposed    17.9
 4 exposed    13.9
 5 exposed    14.1
 6 exposed    18.4
 7 exposed    15.0
 8 exposed    10.3
 9 exposed    11.8
10 exposed    12.5
# … with 110 more rows</code></pre>
</div>
</div>
<p>The <code>oystercatcher</code> data set contains two columns:</p>
<ol type="1">
<li>a <code>site</code> column with information on the amount of shelter of the feeding location</li>
<li>a <code>feeding</code> column containing feeding rates</li>
</ol>
<p>Next, we get some basic descriptive statistics:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># get some basic descriptive statistics</span></span>
<span><span class="va">oystercatcher</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">site</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/get_summary_stats.html">get_summary_stats</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"common"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 11
  site      variable     n   min   max median   iqr  mean    sd    se    ci
  &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 exposed   feeding     40  8.35  18.6   13.9  3.40  13.8  2.44 0.386 0.781
2 partial   feeding     40 10.8   23.0   16.9  2.82  17.1  2.62 0.414 0.838
3 sheltered feeding     40 18.9   28.5   23.2  3.79  23.4  2.42 0.383 0.774</code></pre>
</div>
</div>
<p>Finally, we plot the data by <code>site</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the data</span></span>
<span><span class="va">oystercatcher</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">site</span>, y <span class="op">=</span> <span class="va">feeding</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p>First we read in the data.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load data</span></span>
<span><span class="va">oystercatcher_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"data/CS2-oystercatcher-feeding.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># have a look</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">oystercatcher_r</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     site  feeding
1 exposed 12.17551
2 exposed 13.07392
3 exposed 17.93969
4 exposed 13.89178
5 exposed 14.05166
6 exposed 18.36498</code></pre>
</div>
</div>
<p>The <code>oystercatcher</code> data set contains two columns:</p>
<ol type="1">
<li>a <code>site</code> column with information on the amount of shelter of the feeding location</li>
<li>a <code>feeding</code> column containing feeding rates</li>
</ol>
<p>Next, we get some basic descriptive statistics. We have three groups, so to get the summary statistics by group we do the following:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>          data <span class="op">=</span> <span class="va">oystercatcher_r</span>,</span>
<span>          <span class="va">summary</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       site feeding.Min. feeding.1st Qu. feeding.Median feeding.Mean
1   exposed     8.350801       12.184961      13.946420    13.822899
2   partial    10.795969       15.601927      16.927683    17.081666
3 sheltered    18.856999       21.403028      23.166246    23.355503
  feeding.3rd Qu. feeding.Max.
1       15.581748    18.560404
2       18.416708    23.021250
3       25.197096    28.451252</code></pre>
</div>
</div>
<p>Finally, we plot the data by <code>site</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the data by site</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>        data <span class="op">=</span> <span class="va">oystercatcher_r</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<p>First, we read in the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load the data</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>oystercatcher_py <span class="op">=</span> pd.read_csv(<span class="st">"data/CS2-oystercatcher-feeding.csv"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># and have a look</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>oystercatcher_py.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      site    feeding
0  exposed  12.175506
1  exposed  13.073917
2  exposed  17.939687
3  exposed  13.891783
4  exposed  14.051663</code></pre>
</div>
</div>
<p>The <code>oystercatcher_py</code> data set contains two columns:</p>
<ol type="1">
<li>a <code>site</code> column with information on the amount of shelter of the feeding location</li>
<li>a <code>feeding</code> column containing feeding rates</li>
</ol>
<p>Next, we get some basic descriptive statistics <em>per group</em>. Here we use the <code>pd.groupby()</code> function to group by <code>site</code>. We only want to have summary statistics for the <code>feeding</code> variable, so we specify that as well:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>oystercatcher_py.groupby(<span class="st">"site"</span>)[<span class="st">"feeding"</span>].describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           count       mean       std  ...        50%        75%        max
site                                   ...                                 
exposed     40.0  13.822899  2.441974  ...  13.946420  15.581748  18.560404
partial     40.0  17.081666  2.619906  ...  16.927683  18.416708  23.021250
sheltered   40.0  23.355503  2.419825  ...  23.166246  25.197096  28.451252

[3 rows x 8 columns]</code></pre>
</div>
</div>
<p>Finally, we plot the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  ggplot(oystercatcher_py,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    aes(x <span class="op">=</span> <span class="st">"site"</span>, y <span class="op">=</span> <span class="st">"feeding"</span>))</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> geom_boxplot()</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="614"></p>
</div>
</div>
</div>
</div>
</div>
<p>Looking at the data, there appears to be a noticeable difference in feeding rates between the three sites. We would probably expect a reasonably significant statistical result here.</p>
</section><section id="assumptions" class="level2"><h2 class="anchored" data-anchor-id="assumptions">Assumptions</h2>
<p>To use an ANOVA test, we have to make three assumptions:</p>
<ol type="1">
<li>The parent distributions from which the samples are taken are normally distributed</li>
<li>Each data point in the samples is independent of the others</li>
<li>The parent distributions should have the same variance</li>
</ol>
<p>In a similar way to the two-sample tests we will consider the normality and equality of variance assumptions both using tests and by graphical inspection (and ignore the independence assumption).</p>
<section id="normality" class="level3"><h3 class="anchored" data-anchor-id="normality">Normality</h3>
<p>First we perform a Shapiro-Wilk test on each site separately.</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-3" role="tab" aria-controls="tabset-3-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<p>We take the data, and <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> site:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Shapiro-Wilk test on each site</span></span>
<span><span class="va">oystercatcher</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">site</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/shapiro_test.html">shapiro_test</a></span><span class="op">(</span><span class="va">feeding</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<p>Unstack the data and perform a Shapiro-Wilk test on each group separately.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a new object (a list) that contains the unstacked data</span></span>
<span><span class="va">uns_oystercatcher</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/stack.html">unstack</a></span><span class="op">(</span><span class="va">oystercatcher_r</span>,</span>
<span>                             form <span class="op">=</span> <span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span><span class="op">)</span></span>
<span><span class="co"># have a look at the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">uns_oystercatcher</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   exposed  partial sheltered
1 12.17551 15.20345  23.87176
2 13.07392 16.53239  21.41133
3 17.93969 13.64547  22.46681
4 13.89178 23.02125  23.00843
5 14.05166 20.39774  27.85932
6 18.36498 14.03391  22.09400</code></pre>
</div>
</div>
<p>Next, we perform the Shapiro-Wilk test on each group:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">uns_oystercatcher</span><span class="op">$</span><span class="va">exposed</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  uns_oystercatcher$exposed
W = 0.98859, p-value = 0.953</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">uns_oystercatcher</span><span class="op">$</span><span class="va">partial</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  uns_oystercatcher$partial
W = 0.98791, p-value = 0.9398</code></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">uns_oystercatcher</span><span class="op">$</span><span class="va">sheltered</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  uns_oystercatcher$sheltered
W = 0.97511, p-value = 0.5136</code></pre>
</div>
</div>
</div>
<div id="tabset-3-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-3-tab">
<p>We use the <code>pg.normality()</code> function to calculate the statistic. This requires:</p>
<ul>
<li>the <code>dv</code> dependent variable (<code>feeding</code> in our case)</li>
<li>the <code>group</code> variable (<code>site</code>)</li>
<li>and some data</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pg.normality(dv <span class="op">=</span> <span class="st">"feeding"</span>,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>             group <span class="op">=</span> <span class="st">"site"</span>,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>             data <span class="op">=</span> oystercatcher_py)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  W      pval  normal
exposed    0.988587  0.953037    True
partial    0.987907  0.939830    True
sheltered  0.975106  0.513547    True</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We can see that all three groups appear to be normally distributed which is good.</p>
<p>For ANOVA however, considering each group in turn is often considered quite excessive and, in most cases, it is sufficient to consider the normality of the combined set of <em>residuals</em> from the data. We’ll explain residuals properly in the <a href="#cs3-intro">next session</a> but effectively they are the difference between each data point and its group mean. The residuals can be obtained directly from a linear model fitted to the data.</p>
<p>So, we create a linear model, extract the residuals and check their normality:</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-3" role="tab" aria-controls="tabset-4-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># define the model</span></span>
<span><span class="va">lm_oystercatcher</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">oystercatcher</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the residuals</span></span>
<span><span class="va">resid_oyster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">residuals</a></span><span class="op">(</span><span class="va">lm_oystercatcher</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform Shapiro-Wilk test on residuals</span></span>
<span><span class="va">resid_oyster</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/shapiro_test.html">shapiro_test</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  variable statistic p.value
  &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;
1 .            0.994   0.857</code></pre>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># define the model</span></span>
<span><span class="va">lm_oystercatcher_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>                         data <span class="op">=</span> <span class="va">oystercatcher_r</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the residuals</span></span>
<span><span class="va">resid_oyster_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">residuals</a></span><span class="op">(</span><span class="va">lm_oystercatcher_r</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform Shapiro-Wilk test on residuals</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">resid_oyster_r</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  resid_oyster_r
W = 0.99355, p-value = 0.8571</code></pre>
</div>
</div>
</div>
<div id="tabset-4-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-3-tab">
<p>Unfortunately <code>pingouin</code> does not have a straightforward way of extracting residuals (if you know more, please raise an issue!).</p>
<p>To get our residuals we use <code>statsmodels</code>, a module that provides functions for statistical models. We’ll be using this in upcoming sessions, so you’ll have a head start!</p>
<p>At this point you shouldn’t concern yourself too much with the exact syntax, just run it an have a look.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Technical details (optional)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We need to import a few extra modules. First, we load the <code>statsmodels.api</code> module, which contains an <code>OLS()</code> function (Ordinary Least Squares - the equivalent of the <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code> function in R).</p>
<p>We also import <code>stats.models.formula.api</code> so we can use the formula notation in our linear model. We define the formula as <code>formula= "feeding ~ C(site)"</code> with <code>C</code> conveying that the <code>site</code> variable is a category. Lastly we can <code>.fit()</code> the model.</p>
<p>If you’re familiar with this stuff then you can look at the model itself by running <code>summary(lm_oystercatcher_py)</code>. But we’ll cover all of this in later sessions.</p>
</div>
</div>
</div>
<p>We load the modules, define a linear model, create a <code>fit()</code> and we get the residuals from the linear model fit with <code>.resid</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.api <span class="im">as</span> sm</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.formula.api <span class="im">as</span> smf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a linear model</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> smf.ols(formula<span class="op">=</span> <span class="st">"feeding ~ C(site)"</span>, data <span class="op">=</span> oystercatcher_py)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and get the fitted parameters of the model</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>lm_oystercatcher_py <span class="op">=</span> model.fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the residuals from the model fit</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and perform Shapiro-Wilk test</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>pg.normality(lm_oystercatcher_py.resid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          W     pval  normal
0  0.993546  0.85703    True</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Again, we can see that the combined residuals from all three groups appear to be normally distributed (which is as we would have expected given that they were all normally distributed individually!)</p>
</section><section id="equality-of-variance" class="level3"><h3 class="anchored" data-anchor-id="equality-of-variance">Equality of Variance</h3>
<p>We now test for equality of variance using Bartlett’s test (since we’ve just found that all of the individual groups are normally distributed).</p>
<p>Perform Bartlett’s test on the data:</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-3" role="tab" aria-controls="tabset-5-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># check equality of variance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/bartlett.test.html">bartlett.test</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>              data <span class="op">=</span> <span class="va">oystercatcher</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  feeding by site
Bartlett's K-squared = 0.29598, df = 2, p-value = 0.8624</code></pre>
</div>
</div>
<p>Where the relevant p-value is given on the 3rd line. Here we see that each group appears to have the same variance.</p>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># check equality of variance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/bartlett.test.html">bartlett.test</a></span><span class="op">(</span><span class="va">feeding</span> <span class="op">~</span> <span class="va">site</span>,</span>
<span>              data <span class="op">=</span> <span class="va">oystercatcher</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  feeding by site
Bartlett's K-squared = 0.29598, df = 2, p-value = 0.8624</code></pre>
</div>
</div>
<p>Where the relevant p-value is given on the 3rd line. Here we see that each group appears to have the same variance.</p>
</div>
<div id="tabset-5-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-3-tab">
<p>We use the <code>homoscedasticity()</code> function from <code>pingouin</code> (homoscedasticity is another way of describing equality of variance). The default <code>method</code> is <code>levene</code>, so we need to specify that we want to use <code>bartlett</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>pg.homoscedasticity(dv <span class="op">=</span> <span class="st">"feeding"</span>,</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>                    group <span class="op">=</span> <span class="st">"site"</span>,</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>                    method <span class="op">=</span> <span class="st">"bartlett"</span>,</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>                    data <span class="op">=</span> oystercatcher_py)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 T      pval  equal_var
bartlett  0.295983  0.862439       True</code></pre>
</div>
</div>
<p>Where the relevant p-value is given in the <code>pval</code> column. Here we see that each group appears to have the same variance.</p>
</div>
</div>
</div>
</section><section id="graphical-interpretation-and-diagnostic-plots" class="level3"><h3 class="anchored" data-anchor-id="graphical-interpretation-and-diagnostic-plots">Graphical interpretation and diagnostic plots</h3>
<p>Assessing assumptions via these tests can be cumbersome, but also a bit misleading at times. It reduces the answer to the question “is the assumption met?” to a yes/no, based on some statistic and associated p-value.</p>
<p>This does not convey that things aren’t always so clear-cut and that there is a lot of grey area that we need to navigate. As such, assessing assumptions through graphical means - using diagnostic plots - is often preferred.</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-3" role="tab" aria-controls="tabset-6-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<p>In the first session we already created diagnostic Q-Q plots directly from our data, using <code><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html">stat_qq()</a></code> and <code><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html">stat_qq_line()</a></code>. For more specific plots this becomes a bit cumbersome. There is an option to create ggplot-friendly diagnostic plots, using the <code>ggResidPanel</code> package.</p>
<p>Let’s create the diagnostic plots we’re interested in using <code>ggResidPanel</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lm_oystercatcher</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">resid_panel</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"resid"</span>, <span class="st">"qq"</span>, <span class="st">"ls"</span>, <span class="st">"cookd"</span><span class="op">)</span>,</span>
<span>              smoother <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<ul>
<li>The top left graph plots the <strong>Residuals plot</strong>. If the data are best explained by a linear line then there should be a uniform distribution of points above and below the horizontal blue line (and if there are sufficient points then the red line, which is a smoother line, should be on top of the blue line). This plot looks pretty good.</li>
<li>The top right graph shows the <strong>Q-Q plot</strong> which allows a visual inspection of normality. If the residuals are normally distributed, then the points should lie on the diagonal blue line. This plot looks good.</li>
<li>The bottom left <strong>Location-Scale</strong> graph allows us to investigate whether there is any correlation between the residuals and the predicted values and whether the variance of the residuals changes significantly. If not, then the red line should be horizontal. If there is any correlation or change in variance then the red line will not be horizontal. This plot is fine.</li>
<li>The last graph shows the <strong>Cook’s distance</strong> and tests if any one point has an unnecessarily large effect on the fit. A rule of thumb is that if any value is larger than 1.0, then it might have a large effect on the model. If not, then no point has undue influence. This plot is good. There are different ways to determine the threshold (apart from simply setting it to 1) and in this plot the blue dashed line is at <code>4/n</code>, with <code>n</code> being the number of samples. At this threshold there are some data points that may be influential, but I personally find this threshold rather strict.</li>
</ul>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<p>We can create the default diagnostic plots in base R by simply plotting the <code>lm</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a neat 2x2 window</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># create the diagnostic plots</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">lm_oystercatcher</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># and return the window back to normal</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The second line creates four diagnostic plots.</p>
<ul>
<li>The top left graph plots the <strong>Residuals vs Fitted</strong>. If the data are best explained by a linear line then there should be a uniform distribution of points above and below the horizontal red line. This plot looks pretty good.</li>
<li>The top right graph shows the <strong>Normal Q-Q plot</strong> which allows a visual inspection of normality. If the residuals are normally distributed, then the points should lie on the diagonal dotted line. This plot looks good.</li>
<li>The bottom left <strong>Scale-Location</strong> graph allows us to investigate whether there is any correlation between the residuals and the predicted values and whether the variance of the residuals changes significantly. If not, then the red line should be horizontal. If there is any correlation or change in variance then the red line will not be horizontal. This plot is fine.</li>
<li>The bottom right <strong>Residuals vs Factor Levels</strong> plot shows the residuals for each group (= site). This plot is only displayed if there are equal number of observations in each group and we’ll explain more about this in the next session.</li>
</ul>
</div>
<div id="tabset-6-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-3-tab">
<p>Unfortunately Python doesn’t provide a convenient way of displaying the same diagnostic plots as R does.</p>
<p>I created a function <code>dgplots()</code> (which stands for Diagnostic Plots, very original I know…) that does this for you. All you need to do is create a linear model, get the fit and feed that to the <code>dgplots()</code> function.</p>
<p>You can of course plot the model values yourself by extracting them from the linear model fit, but this should provide a convenient way to avoid that kind of stuff.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>dgplots(lm_oystercatcher_py)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">
<div class="cell-output-display">
<p><img src="images/dgplots/2022_09_20-04:02:37_pm_dgplots.png" class="img-fluid" width="775"></p>
</div>
</div>
<ul>
<li>The top left graph plots the <strong>Residuals plot</strong>. If the data are best explained by a linear line then there should be a uniform distribution of points above and below the horizontal blue line (and if there are sufficient points then the red line, which is a smoother line, should be on top of the blue line). This plot looks pretty good.</li>
<li>The top right graph shows the <strong>Q-Q plot</strong> which allows a visual inspection of normality. If the residuals are normally distributed, then the points should lie on the diagonal blue line. This plot looks good.</li>
<li>The bottom left <strong>Location-Scale</strong> graph allows us to investigate whether there is any correlation between the residuals and the predicted values and whether the variance of the residuals changes significantly. If not, then the red line should be horizontal. If there is any correlation or change in variance then the red line will not be horizontal. This plot is fine.</li>
<li>The last graph shows the <strong>Influential points</strong> and tests if any one point has an unnecessarily large effect on the fit. Here we’re using the Cook’s distance as a measure. A rule of thumb is that if any value is larger than 1.0, then it might have a large effect on the model. If not, then no point has undue influence. This plot is good. There are different ways to determine the threshold (apart from simply setting it to 1) and in this plot the blue dashed line is at <code>4/n</code>, with <code>n</code> being the number of samples. At this threshold there are some data points that may be influential, but I personally find this threshold rather strict.</li>
</ul>
</div>
</div>
</div>
<p>We can see that these graphs are very much in line with what we’ve just looked at using the test, which is reassuring. The groups all appear to have the same spread of data, and the Q-Q plot shows that the assumption of normality is alright.</p>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Assessing assumptions
</div>
</div>
<div class="callout-body-container callout-body">
<p>At this stage, I should point out that I nearly always stick with the graphical method for assessing the assumptions of a test. Assumptions are rarely either completely met or not met and there is always some degree of personal assessment.</p>
<p>Whilst the formal statistical tests (like Shapiro-Wilk) are technically fine, they can often create a false sense of things being absolutely right or wrong in spite of the fact that they themselves are still probabilistic statistical tests. In these exercises we are using both approaches whilst you gain confidence and experience in interpreting the graphical output and whilst it is absolutely fine to use both in the future I would strongly recommend that you don’t rely solely on the statistical tests in isolation.</p>
</div>
</div>
</section></section><section id="implement-and-interpret-the-test" class="level2"><h2 class="anchored" data-anchor-id="implement-and-interpret-the-test">Implement and interpret the test</h2>
<p>Perform an ANOVA test on the data:</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-2" role="tab" aria-controls="tabset-7-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-3" role="tab" aria-controls="tabset-7-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">lm_oystercatcher</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: feeding
           Df  Sum Sq Mean Sq F value    Pr(&gt;F)    
site        2 1878.02  939.01  150.78 &lt; 2.2e-16 ***
Residuals 117  728.63    6.23                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>This takes the linear model (i.e.&nbsp;finds the means of the three groups and calculates a load of intermediary data that we need for the statistical analysis) and stores this information in an R object (which we’ve called <code>lm_oystercatcher</code>, but which you can call what you like).</p>
<p>In the output:</p>
<ul>
<li>The 1st line just tells you the that this is an ANOVA test</li>
<li>The 2nd line tells you what the response variable is (in this case feeding)</li>
<li>The 3rd, 4th and 5th lines are an ANOVA table which contain some useful values:
<ul>
<li>The <code>Df</code> column contains the degrees of freedom values on each row, 2 and 117 (which we can use for the reporting)</li>
<li>The <code>F</code> value column contains the F statistic, 150.78 (which again we’ll need for reporting).</li>
<li>The p-value is 2.2e-16 and is the number directly under the <code>Pr(&gt;F)</code> on the 4th line (to be precise, it is 4.13e-33 but anything smaller than 2.2e-16 gets reported as <code>&lt; 2.2e-16</code>).</li>
<li>The other values in the table (in the <code>Sum Sq</code> and <code>Mean Sq</code>) columns are used to calculate the F statistic itself and we don’t need to know these.</li>
</ul>
</li>
<li>The 6th line has some symbolic codes to represent how big (small) the p-value is; so, a p-value smaller than 0.001 would have a *** symbol next to it (which ours does). Whereas if the p-value was between 0.01 and 0.05 then there would simply be a * character next to it, etc. Thankfully we can all cope with actual numbers and don’t need a short-hand code to determine the reporting of our experiments (please tell me that’s true…!)</li>
</ul>
</div>
<div id="tabset-7-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-2-tab">
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">lm_oystercatcher_r</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: feeding
           Df  Sum Sq Mean Sq F value    Pr(&gt;F)    
site        2 1878.02  939.01  150.78 &lt; 2.2e-16 ***
Residuals 117  728.63    6.23                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>This takes the linear model (i.e.&nbsp;finds the means of the three groups and calculates a load of intermediary data that we need for the statistical analysis) and stores this information in an R object (which we’ve called <code>lm_oystercatcher_r</code>, but which you can call what you like).</p>
<p>In the output:</p>
<ul>
<li>The 1st line just tells you the that this is an ANOVA test</li>
<li>The 2nd line tells you what the response variable is (in this case feeding)</li>
<li>The 3rd, 4th and 5th lines are an ANOVA table which contain some useful values:</li>
<li>The <code>Df</code> column contains the degrees of freedom values on each row, 2 and 117 (which we can use for the reporting)</li>
<li>The <code>F</code> value column contains the F statistic, 150.78 (which again we’ll need for reporting).</li>
<li>The p-value is 2.2e-16 and is the number directly under the <code>Pr(&gt;F)</code> on the 4th line (to be precise, it is 4.13e-33 but anything smaller than 2.2e-16 gets reported as <code>&lt; 2.2e-16</code>).</li>
<li>The other values in the table (in the <code>Sum Sq</code> and <code>Mean Sq</code>) columns are used to calculate the F statistic itself and we don’t need to know these.</li>
<li>The 6th line has some symbolic codes to represent how big (small) the p-value is; so, a p-value smaller than 0.001 would have a *** symbol next to it (which ours does). Whereas if the p-value was between 0.01 and 0.05 then there would simply be a * character next to it, etc. Thankfully we can all cope with actual numbers and don’t need a short-hand code to determine the reporting of our experiments (please tell me that’s true…!)</li>
</ul>
</div>
<div id="tabset-7-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-3-tab">
<p>There are different ways of conducting an ANOVA in Python, with <code>scipy.stats</code> <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.f_oneway.html">proving an option</a>. However, I find that <a href="https://pingouin-stats.org/generated/pingouin.anova.html#pingouin.anova">anova()</a> function in <code>pingouin</code> provides the easiest and most-detailed option to do this.</p>
<p>It takes the following arguments:</p>
<ul>
<li>
<code>dv</code>: dependent variable (response variable; in our case <code>feeding</code>)</li>
<li>
<code>between</code>: between-subject factor (predictor variable; in our case <code>site</code>)</li>
<li>
<code>data</code>: which function doesn’t!?</li>
<li>
<code>detailed</code>: optional <code>True</code> or <code>False</code>, we’re setting it to <code>True</code> because we like to know what we’re doing!</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>pg.anova(dv <span class="op">=</span> <span class="st">"feeding"</span>,</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>         between <span class="op">=</span> <span class="st">"site"</span>,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>         data <span class="op">=</span> oystercatcher_py,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>         detailed <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Source           SS   DF          MS           F         p-unc       np2
0    site  1878.015371    2  939.007685  150.782449  4.128088e-33  0.720473
1  Within   728.625249  117    6.227566         NaN           NaN       NaN</code></pre>
</div>
</div>
<p>This creates a linear model based on the data, <em>i.e</em>. finds the means of the three groups and calculates a load of intermediary data that we need for the statistical analysis.</p>
<p>In the output:</p>
<ul>
<li>
<code>Source</code>: Factor names - in our case these are the different sites (<code>site</code>)</li>
<li>
<code>SS</code>: Sums of squares (we’ll get to that in a bit)</li>
<li>
<code>DF</code>: Degrees of freedom (at the moment only used for reporting)</li>
<li>
<code>MS</code>: Mean squares</li>
<li>
<code>F</code>: Our F-statistic</li>
<li>
<code>p-unc</code>: p-value (<code>unc</code> stands for “uncorrected” - more on multiple testing correction later)</li>
<li>
<code>np2</code>: Partial eta-square effect sizes (more on this later)</li>
</ul>
</div>
</div>
</div>
<p>Again, the p-value is what we’re most interested in here and shows us the probability of getting samples such as ours if the null hypothesis were actually true.</p>
<p>Since the p-value is very small (much smaller than the standard significance level of 0.05) we can say “that it is very unlikely that these three samples came from the same parent distribution” and as such we can reject our null hypothesis and state that:</p>
<blockquote class="blockquote">
<p>A one-way ANOVA showed that the mean feeding rate of oystercatchers differed significantly between locations (F = 150.78, df = 2, 117, p = 4.13e-33).</p>
</blockquote>
</section><section id="post-hoc-testing-tukeys-rank-test" class="level2"><h2 class="anchored" data-anchor-id="post-hoc-testing-tukeys-rank-test">Post-hoc testing (Tukey’s rank test)</h2>
<p>One drawback with using an ANOVA test is that it only tests to see if all of the means are the same, and if we get a significant result using ANOVA then all we can say is that not all of the means are the same, rather than anything about how the pairs of groups differ. For example, consider the following box plot for three samples.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load the data</span></span>
<span><span class="va">tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/CS2-tukey.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tukey</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 60 × 2
   response group  
      &lt;dbl&gt; &lt;chr&gt;  
 1    1.58  sample1
 2    0.380 sample1
 3   -0.997 sample1
 4   -0.771 sample1
 5    0.169 sample1
 6   -0.698 sample1
 7   -0.167 sample1
 8    1.38  sample1
 9   -0.839 sample1
10   -1.05  sample1
# … with 50 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tukey</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">response</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Each group is a random sample of 20 points from a normal distribution with variance 1. Groups 1 and 2 come from a parent population with mean 0 whereas group 3 come from a parent population with mean 2. The data clearly satisfy the assumptions of an ANOVA test.</p>
<section id="read-in-data-and-plot" class="level3"><h3 class="anchored" data-anchor-id="read-in-data-and-plot">Read in data and plot</h3>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-8-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-1" role="tab" aria-controls="tabset-8-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-2" role="tab" aria-controls="tabset-8-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-3" role="tab" aria-controls="tabset-8-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-8-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-8-1-tab">
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load the data</span></span>
<span><span class="va">tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/CS2-tukey.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 60 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): group
dbl (1): response

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># have a look at the data</span></span>
<span><span class="va">tukey</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 60 × 2
   response group  
      &lt;dbl&gt; &lt;chr&gt;  
 1    1.58  sample1
 2    0.380 sample1
 3   -0.997 sample1
 4   -0.771 sample1
 5    0.169 sample1
 6   -0.698 sample1
 7   -0.167 sample1
 8    1.38  sample1
 9   -0.839 sample1
10   -1.05  sample1
# … with 50 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the data</span></span>
<span><span class="va">tukey</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">response</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-8-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-2-tab">
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"data/CS2-tukey.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">response</span> <span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">tukey</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-8-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-3-tab">

</div>
</div>
</div>
</section><section id="test-for-a-significant-difference-in-group-means" class="level3"><h3 class="anchored" data-anchor-id="test-for-a-significant-difference-in-group-means">Test for a significant difference in group means</h3>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-9-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-1" role="tab" aria-controls="tabset-9-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-2" role="tab" aria-controls="tabset-9-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-3" role="tab" aria-controls="tabset-9-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-9-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-9-1-tab">
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a linear model</span></span>
<span><span class="va">lm_tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">response</span> <span class="op">~</span> <span class="va">group</span>,</span>
<span>               data <span class="op">=</span> <span class="va">tukey</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform an ANOVA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">lm_tukey</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: response
          Df Sum Sq Mean Sq F value    Pr(&gt;F)    
group      2 33.850 16.9250   20.16 2.392e-07 ***
Residuals 57 47.854  0.8395                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
<div id="tabset-9-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-2-tab">
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lm_tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">response</span> <span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">tukey</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">lm_tukey</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: response
          Df Sum Sq Mean Sq F value    Pr(&gt;F)    
group      2 33.850 16.9250   20.16 2.392e-07 ***
Residuals 57 47.854  0.8395                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
<div id="tabset-9-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-3-tab">

</div>
</div>
</div>
<p>Here we have a p-value of 2.39 <span class="math inline">\(\times\)</span> 10<sup>-7</sup> and so the test has very conclusively rejected the hypothesis that all means are equal.</p>
<p>However, this was not due to all of the sample means being different, but rather just because one of the groups is very different from the others. In order to drill down and investigate this further we use a new test called <strong>Tukey’s range test</strong> (or <strong>Tukey’s honest significant difference test</strong> – this always makes me think of some terrible cowboy/western dialogue).</p>
<p>This will compare all of the groups in a pairwise fashion and reports on whether a significant difference exists.</p>
</section><section id="performing-tukeys-test" class="level3"><h3 class="anchored" data-anchor-id="performing-tukeys-test">Performing Tukey’s test</h3>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-10-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-1" role="tab" aria-controls="tabset-10-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-2" role="tab" aria-controls="tabset-10-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-3" role="tab" aria-controls="tabset-10-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-10-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-10-1-tab">
<div class="cell">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform Tukey's range test on linear model</span></span>
<span><span class="va">lm_tukey</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/tukey_hsd.html">tukey_hsd</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 9
  term  group1  group2  null.value estimate conf.low conf.high       p.adj
* &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;       &lt;dbl&gt;
1 group sample1 sample2          0    0.304   -0.393      1.00 0.55       
2 group sample1 sample3          0    1.72     1.03       2.42 0.000000522
3 group sample2 sample3          0    1.42     0.722      2.12 0.0000246  
# … with 1 more variable: p.adj.signif &lt;chr&gt;</code></pre>
</div>
</div>
<p>The <code><a href="https://rpkgs.datanovia.com/rstatix/reference/tukey_hsd.html">tukey_hsd()</a></code> function takes our linear model (<code>lm_tukey</code>) as its input. The output is a pair-by-pair comparison between the different groups (samples 1 to 3). We are interested in the <code>p.adj</code> column, which gives us the adjusted p-value. The null hypothesis in each case is that there is no difference in the mean between the two groups.</p>
</div>
<div id="tabset-10-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-2-tab">
<div class="cell">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aov_tukey</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aov.html">aov</a></span><span class="op">(</span><span class="va">response</span> <span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">tukey</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD</a></span><span class="op">(</span><span class="va">aov_tukey</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = response ~ group, data = tukey)

$group
                     diff        lwr      upr     p adj
sample2-sample1 0.3037563 -0.3934982 1.001011 0.5498005
sample3-sample1 1.7233591  1.0261047 2.420614 0.0000005
sample3-sample2 1.4196028  0.7223484 2.116857 0.0000246</code></pre>
</div>
</div>
<p>The first argument repeats our ANOVA using a different function <code><a href="https://rdrr.io/r/stats/aov.html">aov()</a></code>. We store the output of this function in an R object called <code>aov_tukey</code>. Note that the <code><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD()</a></code> function takes the output of the <code><a href="https://rdrr.io/r/stats/aov.html">aov()</a></code> function as its argument and not the raw data.</p>
<p>The bottom three lines contain the information that we want. The final column of each (entitled <code>p adj</code>) is the p-value that we’re looking for. The null hypothesis in each case is that there is no difference in the mean between the two groups.</p>
</div>
<div id="tabset-10-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-3-tab">

</div>
</div>
</div>
<p>As we can see that there isn’t a significant difference between <code>sample1</code> and <code>sample2</code> but the 2nd and 3rd rows show that there is a significant difference between <code>sample1</code> and <code>sample3</code>, as well as <code>sample2</code> and <code>sample3</code>. This matches with what we expected based on the box plot.</p>
</section><section id="assumptions-1" class="level3"><h3 class="anchored" data-anchor-id="assumptions-1">Assumptions</h3>
<p>When to use Tukey’s range test is a matter of debate (strangely enough a lot of statistical analysis techniques are currently matters of opinion rather than mathematical fact – it does explain a little why this whole field appears so bloody confusing!)</p>
<ul>
<li>Some people claim that we should only perform Tukey’s range test (or any other post-hoc tests) if the preceding ANOVA test showed that there was a significant difference between the groups and that if the ANOVA test had not shown any significant differences between groups then we would have to stop there.</li>
<li>Other people say that this is rubbish and we can do whatever we like as long as we tell people what we did.</li>
</ul>
<p>The background to this is rather involved but one of the reasons for this debate is to prevent so-called <strong>data-dredging</strong> or <strong>p-hacking</strong>. This is where scientists/analysts are so fixated on getting a “significant” result that they perform a huge variety of statistical techniques until they find one that shows that their data is significant (this was a particular problem in psychological studies for while – not to point fingers though, they are working hard to sort their stuff out. Kudos!).</p>
<p>Whether you should use post-hoc testing or not will depend on your experimental design and the questions that you’re attempting to answer.</p>
<p>Tukey’s range test, when we decide to use it, requires the same three assumptions as an ANOVA test:</p>
<ol type="1">
<li>Normality of distributions</li>
<li>Equality of variance between groups</li>
<li>Independence of observations</li>
</ol></section></section><section id="exercise-lobster-weight" class="level2"><h2 class="anchored" data-anchor-id="exercise-lobster-weight">Exercise: Lobster weight</h2>
<p>Juvenile lobsters in aquaculture were grown on three different diets (fresh mussels, semi-dry pellets and dry flakes). After nine weeks, their wet weight was:</p>
<div class="cell">
<div class="cell-output-display">

<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:right;"> Mussels </th>
   <th style="text-align:right;"> Pellets </th>
   <th style="text-align:right;"> Flakes </th>
  </tr></thead>
<tbody>
<tr>
<td style="text-align:right;"> 151.6 </td>
   <td style="text-align:right;"> 117.7 </td>
   <td style="text-align:right;"> 101.8 </td>
  </tr>
<tr>
<td style="text-align:right;"> 132.1 </td>
   <td style="text-align:right;"> 110.8 </td>
   <td style="text-align:right;"> 102.9 </td>
  </tr>
<tr>
<td style="text-align:right;"> 104.2 </td>
   <td style="text-align:right;"> 128.6 </td>
   <td style="text-align:right;"> 90.4 </td>
  </tr>
<tr>
<td style="text-align:right;"> 153.5 </td>
   <td style="text-align:right;"> 110.1 </td>
   <td style="text-align:right;"> 132.8 </td>
  </tr>
<tr>
<td style="text-align:right;"> 132.0 </td>
   <td style="text-align:right;"> 175.2 </td>
   <td style="text-align:right;"> 129.3 </td>
  </tr>
<tr>
<td style="text-align:right;"> 119.0 </td>
   <td style="text-align:right;"> NA </td>
   <td style="text-align:right;"> 129.4 </td>
  </tr>
<tr>
<td style="text-align:right;"> 161.9 </td>
   <td style="text-align:right;"> NA </td>
   <td style="text-align:right;"> NA </td>
  </tr>
</tbody>
</table>
</div>
</div>
<p>Is there any evidence that diet affects the growth rate of lobsters?</p>
<ol type="1">
<li>Write down the null and alternative hypotheses</li>
<li>Import the data into R from <code>data/CS2-lobsters.csv</code>
</li>
<li>Summarise and visualise the data</li>
<li>Check the assumptions using appropriate tests and graphical analyses</li>
<li>Perform an ANOVA test</li>
<li>Write down a sentence that summarise the results that you have found</li>
<li>Perform a post-hoc test and report the findings</li>
</ol>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="hypotheses" class="level3"><h3 class="anchored" data-anchor-id="hypotheses">Hypotheses</h3>
<p><span class="math inline">\(H_0\)</span> : all means are equal</p>
<p><span class="math inline">\(H_1\)</span> : not all means are equal</p>
</section><section id="import-data-summarise-and-visualise" class="level3"><h3 class="anchored" data-anchor-id="import-data-summarise-and-visualise">Import Data, summarise and visualise</h3>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-11-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-1" role="tab" aria-controls="tabset-11-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-2" role="tab" aria-controls="tabset-11-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-3" role="tab" aria-controls="tabset-11-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-11-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-11-1-tab">
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load the data</span></span>
<span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/CS2-lobsters.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at the data</span></span>
<span><span class="va">lobsters</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 18 × 3
      id weight diet   
   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
 1     1  152.  Mussels
 2     2  132.  Mussels
 3     3  104.  Mussels
 4     4  154.  Mussels
 5     5  132   Mussels
 6     6  119   Mussels
 7     7  162.  Mussels
 8     8  118.  Pellets
 9     9  111.  Pellets
10    10  129.  Pellets
11    11  110.  Pellets
12    12  175.  Pellets
13    13  102.  Flakes 
14    14  103.  Flakes 
15    15   90.4 Flakes 
16    16  133.  Flakes 
17    17  129.  Flakes 
18    18  129.  Flakes </code></pre>
</div>
</div>
<p>The data have a unique <code>id</code> column, which we don’t need any summary statistics for, so we deselect it:</p>
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create some summary statistics</span></span>
<span><span class="va">lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">diet</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/get_summary_stats.html">get_summary_stats</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"common"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 11
  diet    variable     n   min   max median   iqr  mean    sd    se    ci
  &lt;chr&gt;   &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 Flakes  weight       6  90.4  133.   116.  27.3  114.  18.2  7.42  19.1
2 Mussels weight       7 104.   162.   132.  27.0  136.  20.6  7.79  19.1
3 Pellets weight       5 110.   175.   118.  17.8  128.  27.2 12.1   33.7</code></pre>
</div>
</div>
<p>Next, we visualise the data:</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">diet</span>, y <span class="op">=</span> <span class="va">weight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-11-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-2-tab">
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"data/CS2-lobsters.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the data and see what we can see.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>   id weight    diet
1   1  151.6 Mussels
2   2  132.1 Mussels
3   3  104.2 Mussels
4   4  153.5 Mussels
5   5  132.0 Mussels
6   6  119.0 Mussels
7   7  161.9 Mussels
8   8  117.7 Pellets
9   9  110.8 Pellets
10 10  128.6 Pellets
11 11  110.1 Pellets
12 12  175.2 Pellets
13 13  101.8  Flakes
14 14  102.9  Flakes
15 15   90.4  Flakes
16 16  132.8  Flakes
17 17  129.3  Flakes
18 18  129.4  Flakes</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span>, <span class="va">summary</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     diet weight.Min. weight.1st Qu. weight.Median weight.Mean weight.3rd Qu.
1  Flakes     90.4000       102.0750      116.1000    114.4333       129.3750
2 Mussels    104.2000       125.5000      132.1000    136.3286       152.5500
3 Pellets    110.1000       110.8000      117.7000    128.4800       128.6000
  weight.Max.
1    132.8000
2    161.9000
3    175.2000</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-11-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-3-tab">

</div>
</div>
</div>
<p>As always we use the plot and summary to assess three things:</p>
<ol type="1">
<li>Did we load the data in properly?</li>
</ol>
<ul>
<li>We see three groups with reasonable values. There aren’t any data points that are obviously wrong (negative, zero or massively big) and we have the right number of groups. So it looks as if we didn’t do anything obviously wrong.</li>
</ul>
<ol start="2" type="1">
<li>What do we expect as a result of a statistical test?</li>
</ol>
<ul>
<li>Whilst the <code>Mussels</code> group does look higher than the other two groups, <code>Pellets</code> and <code>Flakes</code> appear almost identical in terms of average values, and there’s quite a bit of overlap with the <code>Mussels</code> group. A non-significant result is the most likely answer, and I would be surprised to see a significant p-value - especially given the small sample size that we have here.</li>
</ul>
<ol start="3" type="1">
<li>What do we think about assumptions?</li>
</ol>
<ul>
<li>The groups appear mainly symmetric (although <code>Pellets</code> is a bit weird) and so we’re not immediately massively worried about lack of normality. Again, <code>Flakes</code> and <code>Mussels</code> appear to have very similar variances but it’s a bit hard to decide what’s going on with <code>Pellets.</code> It’s hard to say what’s going on with the assumptions and so I’ll wait and see what the other tests say.</li>
</ul></section><section id="explore-assumptions" class="level3"><h3 class="anchored" data-anchor-id="explore-assumptions">Explore Assumptions</h3>
<p>We’ll explore the assumption of normality and equality of variance, assuming that the data are independent.</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-12-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-1" role="tab" aria-controls="tabset-12-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-12-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-2" role="tab" aria-controls="tabset-12-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-12-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-3" role="tab" aria-controls="tabset-12-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-12-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-12-1-tab">
<p><strong>Normality</strong></p>
<p>We’ll be really thorough here and consider the normality of each group separately and jointly using the Shapiro-Wilk test, as well as looking at the Q-Q plot. In reality, and after these examples , we’ll only use the Q-Q plot.</p>
<p>First, we perform the Shapiro-Wilk test on the individual groups:</p>
<div class="cell">
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Shapiro-Wilk on lobster groups</span></span>
<span><span class="va">lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">diet</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/shapiro_test.html">shapiro_test</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  diet    variable statistic      p
  &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;
1 Flakes  weight       0.844 0.140 
2 Mussels weight       0.948 0.710 
3 Pellets weight       0.767 0.0425</code></pre>
</div>
</div>
<p><code>Flakes</code> and <code>Mussels</code> are fine, but, as we suspected from earlier, <code>Pellets</code> appears to have a marginally significant Normality test result.</p>
<p>Let’s look at the Shapiro-Wilk test for all of the data together:</p>
<div class="cell">
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a linear model</span></span>
<span><span class="va">lm_lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>,</span>
<span>                  data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the residuals</span></span>
<span><span class="va">resid_lobsters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">residuals</a></span><span class="op">(</span><span class="va">lm_lobsters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># and perform the Shapiro-Wilk test on the residuals</span></span>
<span><span class="va">resid_lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/shapiro_test.html">shapiro_test</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  variable statistic p.value
  &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt;
1 .            0.948   0.391</code></pre>
</div>
</div>
<p>This on the other hand says that everything is fine. Let’s look at Q-Q plot.</p>
<div class="cell">
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Q-Q plots</span></span>
<span><span class="va">lm_lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">resid_panel</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="st">"qq"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here, I’ve used an extra argument to the normal diagnostic plots call. The default option is to plot 4 diagnostic plots. You can tell <code><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">resid_panel()</a></code> to only plot a specific one, using the <code>plots =</code> arguments. If you want to know more about this have a look at the <a href="https://goodekat.github.io/ggResidpanel-tutorial/tutorial.html#overview">help documentation</a> or by using <code><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">?resid_panel</a></code>.</p>
<p>The Q-Q plot looks OK, not perfect, but more than good enough for us to have confidence in the normality of the data.</p>
<p>Overall, I’d be happy that the assumption of normality has been adequately well met here. The suggested lack of normality in the <code>Pellets</code> was only just significant and we have to take into account that there are only 5 data points in that group. If there had been a lot more points in that group, or if the Q-Q plot was considerably worse then I wouldn’t be confident.</p>
<p><strong>Equality of Variance</strong></p>
<p>We’ll consider the Bartlett test and we’ll look at some diagnostic plots too.</p>
<div class="cell">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform Bartlett's test</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/bartlett.test.html">bartlett.test</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>,</span>
<span>              data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  weight by diet
Bartlett's K-squared = 0.71273, df = 2, p-value = 0.7002</code></pre>
</div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the residuals and scale-location plots</span></span>
<span><span class="va">lm_lobsters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://goodekat.github.io/ggResidpanel/reference/resid_panel.html">resid_panel</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"resid"</span>, <span class="st">"ls"</span><span class="op">)</span>,</span>
<span>              smoother <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In the above code I’ve specified which diagnostic plots I wanted. I have also added a loess smoother line (<code>smoother = TRUE</code>) to the plots</p>
<ol type="1">
<li>The Residuals Plot. What we’re looking for there is that the points are evenly spread on either side of the line. Looks good.</li>
<li>The Location-Scale Plot (this is displayed by default in base R’s diagnostic plots). Here we’re looking at the red line. If that line is more or less horizontal, then the equality of variance assumption is met.</li>
</ol>
<p>Here all three methods agree that there aren’t any issues with equality of variance:</p>
<ul>
<li>the Bartlett test p-value is large and non-significant</li>
<li>the spread of points in all three groups in the residuals vs fitted graph are roughly the same</li>
<li>the red line in the scale-location graph is pretty horizontal</li>
</ul>
<p>Overall, this assumption is pretty well met.</p>
</div>
<div id="tabset-12-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-12-2-tab">
<p><strong>Normality</strong></p>
<p>We’ll be really thorough here and consider the normality of each group separately and jointly using the Shapiro-Wilk test, as well as looking at the Q-Q plot. In reality, and after these examples , we’ll only use the Q-Q plot.</p>
<p>We’ll need to unstack the data to use the Shapiro-Wilk test on the individual groups:</p>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lobst_uns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/stack.html">unstack</a></span><span class="op">(</span><span class="va">lobsters</span>, <span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">lobst_uns</span><span class="op">$</span><span class="va">Flakes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  lobst_uns$Flakes
W = 0.84368, p-value = 0.1398</code></pre>
</div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">lobst_uns</span><span class="op">$</span><span class="va">Mussels</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  lobst_uns$Mussels
W = 0.94784, p-value = 0.71</code></pre>
</div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">lobst_uns</span><span class="op">$</span><span class="va">Pellets</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  lobst_uns$Pellets
W = 0.76706, p-value = 0.0425</code></pre>
</div>
</div>
<p><code>Flakes</code> and <code>Mussels</code> are fine, but, as we suspected from earlier, <code>Pellets</code> appears to have a marginally significant Normality test result.</p>
<p>Let’s look at the Shapiro-Wilk test for all of the data together:</p>
<div class="cell">
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">resid_lobst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">residuals</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html">shapiro.test</a></span><span class="op">(</span><span class="va">resid_lobst</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  resid_lobst
W = 0.94779, p-value = 0.3914</code></pre>
</div>
</div>
<p>This on the other hand says that everything is fine. Let’s look at the Q-Q-plot:</p>
<div class="cell">
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span> , data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span>,</span>
<span>     which <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here, I’ve used an extra argument to the normal diagnostic plots call. The default option is to plot 4 diagnostic plots, but you can tell R to only plot a specific one. (If you want to know more about this have a look at the <code>plot.lm</code> help documentation using <code><a href="https://rdrr.io/r/stats/plot.lm.html">?plot.lm</a></code>). I’ve asked R to only plot the Q-Q plot with the <code>which = 2</code> argument.</p>
<p>The Q-Q plot looks OK, not perfect, but more than good enough for us to have confidence in the normality of the data.</p>
<p>Overall, I’d be happy that the assumption of normality has been adequately well met here. The suggested lack of normality in the <code>Pellets</code> was only just significant and we have to take into account that there are only 5 data points in that group. If there had been a lot more points in that group, or if the Q-Q plot was considerably worse then I wouldn’t be confident.</p>
<p><strong>Equality of Variance</strong></p>
<p>We’ll consider the Bartlett test and we’ll look at some diagnostic plots too.</p>
<div class="cell">
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/bartlett.test.html">bartlett.test</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  weight by diet
Bartlett's K-squared = 0.71273, df = 2, p-value = 0.7002</code></pre>
</div>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span>,</span>
<span>     which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="cs2_practical_anova_files/figure-html/unnamed-chunk-58-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In the above code, I’ve used the same trick as before with the <code>which</code> argument to only plot the two diagnostic plots that relate to equality of variance (residuals vs fitted and scale-location).</p>
<p>Here all three methods agree that there aren’t any issues with equality of variance:</p>
<ul>
<li>the Bartlett test p-value is large and non-significant</li>
<li>the spread of points in all three groups in the residuals vs fitted graph are roughly the same</li>
<li>the red line in the scale-location graph is pretty horizontal</li>
</ul>
<p>Overall, this assumption is pretty well met.</p>
</div>
<div id="tabset-12-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-12-3-tab">

</div>
</div>
</div>
</section><section id="carry-out-one-way-anova" class="level3"><h3 class="anchored" data-anchor-id="carry-out-one-way-anova">Carry out one-way ANOVA</h3>
<p>With our assumptions of normality and equality of variance met we can be confident that a one-way ANOVA is an appropriate test.</p>
<div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-13-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-1" role="tab" aria-controls="tabset-13-1" aria-selected="true">tidyverse</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-13-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-2" role="tab" aria-controls="tabset-13-2" aria-selected="false">R</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-13-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-3" role="tab" aria-controls="tabset-13-3" aria-selected="false">Python</a></li>
</ul>
<div class="tab-content" data-group="language">
<div id="tabset-13-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-13-1-tab">
<div class="cell">
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">lm_lobsters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: weight
          Df Sum Sq Mean Sq F value Pr(&gt;F)
diet       2 1567.2  783.61  1.6432 0.2263
Residuals 15 7153.1  476.87               </code></pre>
</div>
</div>
</div>
<div id="tabset-13-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-13-2-tab">
<div class="cell">
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">~</span> <span class="va">diet</span>, data <span class="op">=</span> <span class="va">lobsters</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: weight
          Df Sum Sq Mean Sq F value Pr(&gt;F)
diet       2 1567.2  783.61  1.6432 0.2263
Residuals 15 7153.1  476.87               </code></pre>
</div>
</div>
</div>
<div id="tabset-13-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-13-3-tab">

</div>
</div>
</div>
<blockquote class="blockquote">
<p>A one-way ANOVA test indicated that the mean weight of juvenile lobsters did not differ significantly between diets (F = 1.64, df = 2,15, p = 0.23).</p>
</blockquote>
</section><section id="post-hoc-testing-with-tukey" class="level3"><h3 class="anchored" data-anchor-id="post-hoc-testing-with-tukey">Post-hoc testing with Tukey</h3>
<p>In this case we did not find any significant differences between the different diets. So that is a good time for me to reiterate that carrying out the post-hoc test after getting a non-significant result with ANOVA is something that you have to think very carefully about and it all depends on what your research question it.</p>
<p>If your research question was:</p>
<blockquote class="blockquote">
<p>Does diet affect lobster weight?</p>
</blockquote>
<p>or</p>
<blockquote class="blockquote">
<p>Is there any effect of diet on lobster weight?_</p>
</blockquote>
<p>Then when we got the non-significant result from the ANOVA test we should have just stopped there as we have our answer. Going digging for “significant” results by running more tests is a main factor that contributes towards lack of reproducibility in research.</p>
<p>If on the other hand your research question was:</p>
<blockquote class="blockquote">
<p>Are any specific diets better or worse for lobster weight than others?</p>
</blockquote>
<p>Then we should probably have just skipped the one-way ANOVA test entirely and just jumped straight in with the Tukey’s range test. The important point here is that the result of the one-way ANOVA test doesn’t stop you from carrying out the Tukey test - but it’s up to you to decide whether it is sensible.</p>
</section>
</div>
</div>
</div>
</section><section id="key-points" class="level2"><h2 class="anchored" data-anchor-id="key-points">Key points</h2>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>We use an ANOVA to test if there is a difference in means between multiple continuous response variables</li>
<li>We check assumptions with diagnostic plots and check if the residuals are normally distributed</li>
<li>We use post-hoc testing to check for significant differences between the group means, for example using Tukey’s range test</li>
</ul>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../materials/cs1_practical_two-sample-paired-wilcoxon.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Wilcoxon signed rank test</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/cs2_practical_kruskal-wallis.html" class="pagination-link">
        <span class="nav-page-text">Kruskal-Wallis</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
      <div class="nav-footer-center">Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</div>
  </div>
</footer>


</body></html>